File: test/language/templates/cpp98/ChoiceTemplatedSelectorTest.cpp
#include "gtest/gtest.h"

#include "templates/choice_templated_selector/ChoiceTemplatedSelector.h"

namespace templates
{
namespace choice_templated_selector
{

TEST(ChoiceTemplatedSelectorTest, readWrite)
{
    ChoiceTemplatedSelector choiceTemplatedSelector;
    choiceTemplatedSelector.setSelector16(0);
    choiceTemplatedSelector.setSelector32(2);
    TemplatedChoice_uint16& uint16Choice = choiceTemplatedSelector.getUint16Choice();
    uint16Choice.setUint16Field(42);
    TemplatedChoice_uint32& uint32Choice = choiceTemplatedSelector.getUint32Choice();
    uint32Choice.setStringField("string");

    zserio::BitStreamWriter writer;
    choiceTemplatedSelector.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    ChoiceTemplatedSelector readChoiceTemplatedSelector(reader);

    ASSERT_TRUE(choiceTemplatedSelector == readChoiceTemplatedSelector);
}

} // namespace choice_templated_selector
} // namespace templates

File: test/language/templates/java/templates/ChoiceTemplatedSelectorTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.choice_templated_selector.ChoiceTemplatedSelector;
import templates.choice_templated_selector.TemplatedChoice_uint32_Shift32;
import templates.choice_templated_selector.TemplatedChoice_uint16_Shift16;

public class ChoiceTemplatedSelectorTest
{
    @Test
    public void readWrite() throws IOException
    {
        final int selector16 = 0;
        final int selector32 = 1;
        final ChoiceTemplatedSelector choiceTemplatedSelector = new ChoiceTemplatedSelector();
        choiceTemplatedSelector.setSelector16(selector16);
        choiceTemplatedSelector.setSelector32(selector32);

        final TemplatedChoice_uint16_Shift16 uint16Choice = new TemplatedChoice_uint16_Shift16(selector16);
        uint16Choice.setUint16Field(42);
        choiceTemplatedSelector.setUint16Choice(uint16Choice);

        final TemplatedChoice_uint32_Shift32 uint32Choice = new TemplatedChoice_uint32_Shift32(selector32);
        uint32Choice.setStringField("string");
        choiceTemplatedSelector.setUint32Choice(uint32Choice);

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        choiceTemplatedSelector.write(writer);

        final BitStreamReader reader = new ByteArrayBitStreamReader(
                writer.toByteArray(), writer.getBitPosition());
        ChoiceTemplatedSelector readChoiceTemplatedSelector = new ChoiceTemplatedSelector(reader);
        assertTrue(choiceTemplatedSelector.equals(readChoiceTemplatedSelector));
    }
}
